### PR TITLE
Use babel-eslint and update eslint

### LINKED ISSUE #3958: Upgrade eslint, use babel-eslint
We're a few versions back, it would be good to get all upgraded. There have been a few changes to rules which are a bit annoying so we'll probably have to tweak the configs / make a few changes to code.

cc @pipeep

### PR DESCRIPTION
**Pull request by PiPeep**
Fixes #3958.

-   Removes esprima-fb dependency
-   Tightens up eslintrc with some minor rules we were pretty-much following anyways.
-   Adds pretty colors to the `grunt lint` output
-   Breaks `block-scoped-var` :cry:

@hzoo, insight into/thoughts on the ~~`no-empty`~~ and `blocked-scoped-var` rules would be helpful.

### CODE REVIEW

**Comment by spicyj** [comment]
It would be nice if we didn't need to put comments in our empty `catch` blocks.

**Comment by PiPeep** [comment]
@spicyj, @zpao seemed iffy on it too. I'll go ahead and revert that bit.

**Comment by hzoo** [comment]
Ok I'll investigate `block-scoped-var` issues - will probably need to update `bable-eslint`.

File: .eslintrc [hunk 1]
```diff
@@ -46,6 +48,8 @@ rules:
   no-use-before-define: 0
   # We do this in a few places to align values
   key-spacing: 0
+  # https://github.com/facebook/react/pull/3968
+  no-empty: 0
 
-  # DISABLED. These currently cause errors when running.
-  no-multi-spaces: 0
+  # BROKEN. We'd like to turn these back on.
+  block-scoped-var: 0 # causes a ton of noise, eslint is too picky?
```
Comment (zpao) [inline .eslintrc]: Nit: comment above

File: .eslintrc [hunk 2]
```diff
@@ -17,16 +17,18 @@ rules:
   strict: 2
   # We actually have a transform to support this and we fix this for bundled
   # releases but not for the npm package, so enforce it strictly
-  no-comma-dangle: 2
+  comma-dangle: [2, never]
```
Comment (zpao) [inline .eslintrc]: In a followup, let's stop enforcing this (and maybe do the opposite). This does get transformed correctly now.

File: src/renderers/dom/client/ReactBrowserEventEmitter.js [hunk 3]
```diff
@@ -212,8 +212,8 @@ var ReactBrowserEventEmitter = assign({}, ReactEventEmitterMixin, {
   listenTo: function(registrationName, contentDocumentHandle) {
     var mountAt = contentDocumentHandle;
     var isListening = getListeningForDocument(mountAt);
-    var dependencies = EventPluginRegistry.
-      registrationNameDependencies[registrationName];
+    var dependencies = EventPluginRegistry
+      .registrationNameDependencies[registrationName];
```
Comment (zpao) [inline src/renderers/dom/client/ReactBrowserEventEmitter.js]: We'll prefer a newline before splitting like this (except in cases where we're chaining) if it fits in 80. Let's just do that.

```js
    var dependencies = 
      EventPluginRegistry.registrationNameDependencies[registrationName];
```

**Comment by hzoo** [comment]
Ok regarding `block-scoped-var`, it actually seems to be an issue when `ecmaVersion = 6` for escope/eslint (`babel-eslint` makes it always 6). I can get the same issues without `babel-eslint` just adding

```yaml
# parser: babel-eslint
ecmaFeatures:
  modules: true # or another es6 feature
```
Still looking but probably an eslint issue?

Example is `/src/shared/utils/traverseAllChildren.js`

54:13 `warning  "wrapUserProvidedKey" used outside of binding context  block-scoped-var`

**Comment by hzoo** [comment]
So when you enable `block-scoped-var`.
> With block scoping, a variable cannot be used before it is defined. You give up function declaration hoisting when you enable that rule.

So the fix would be to either remove that rule or move the functions declarations higher up in the file before they are called (it seems like that would be a lot)

**Comment by PiPeep** [comment]
@hzoo, But does "cannot be used before it is defined" include

```js
function foo() {
  console.log(a);
}
var a = 3;
foo();
```

I think it's breaking on that type of pattern when it shouldn't, as `a` isn't actually used until `foo` is called (I believe this is a case of late binding, not hoisting).

Regardless, it seems like `block-scoped-var` is duplicating functionality of `no-use-before-define`.

If there's no easy fix, we can leave it disabled for now, and maybe add a custom rule later, or fix all the ~100 places in our code it finds.

**Comment by hzoo** [comment]
@PiPeep I just ran that code snippet and it does error unless you move `var a = 3` above it.
(And this is only when `ecmaFeatures: modules: true` though). @michaelficarra?

And wouldn't need the rule when all var's become let/const right?

**Comment by zpao** [comment]
Let's disable the rule for now. It is interesting that the same code that's failing here with babel-eslint does not fail when using esprima-fb (nor espree IIRC).

File: .eslintrc [hunk 4]
```diff
@@ -46,6 +48,9 @@ rules:
   no-use-before-define: 0
   # We do this in a few places to align values
   key-spacing: 0
+  # https://github.com/facebook/react/pull/3968
```
Comment (zpao) [inline .eslintrc]: Nit: can you actually mention why it's disabled and not just link to a PR which has a lot of unrelated discussion.

**Comment by hzoo** [comment]
@zpao When I add to the `.eslintrc`
```yaml
ecmaFeatures:
  modules: true
```
it does fail on all the parsers.

**Comment by zpao** [comment]
:thumbsup: merge when ready

**Comment by PiPeep** [comment]
Rebased and squashed onto master.

**Comment by zpao** [comment]
Note: pushed with 58fb322e211614ffb4fc6a8228e96e93c5ea65dc

**Pull request closed without merging**